pkg/localstorev2/internal/chunkstore/txchunkstore.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2023 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package chunkstore
+
+import (
+	"context"
+	"sync"
+
+	"github.com/ethersphere/bee/pkg/sharky"
+	storage "github.com/ethersphere/bee/pkg/storagev2"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/hashicorp/go-multierror"
+)
+
+type txSharky struct {
+	Sharky
+
+	opsMu         sync.Mutex
+	releaseOps    []sharky.Location
+	committedLocs []sharky.Location
+}
+
+func (t *txSharky) Write(ctx context.Context, data []byte) (sharky.Location, error) {
+	loc, err := t.Sharky.Write(ctx, data)
+	if err == nil {
+		t.opsMu.Lock()
+		t.committedLocs = append(t.committedLocs, loc)
+		t.opsMu.Unlock()
+	}
+	return loc, err
+}
+
+func (t *txSharky) Release(_ context.Context, loc sharky.Location) error {
+	t.opsMu.Lock()
+	defer t.opsMu.Unlock()
+
+	t.releaseOps = append(t.releaseOps, loc)
+
+	return nil
+}
+
+func (t *txSharky) newTx() *txSharky {
+	return &txSharky{Sharky: t.Sharky}
+}
+
+type txChunkStoreWrapper struct {
+	*storage.TxChunkStoreBase
+
+	txStore storage.TxStore
+	sharky  *txSharky
+	doneMu  sync.Mutex
+}
+
+func (t *txChunkStoreWrapper) Put(ctx context.Context, chunk swarm.Chunk) error {
+	return t.TxChunkStoreBase.Put(ctx, chunk)
+}
+
+func (t *txChunkStoreWrapper) Delete(ctx context.Context, address swarm.Address) error {
+	return t.TxChunkStoreBase.Delete(ctx, address)
+}
+
+func (t *txChunkStoreWrapper) Commit() error {
+	t.doneMu.Lock()
```

Comment:
Not quite sure why this lock is needed here. The same goes for the Rollback method.

Reply:
So the problem I see is that the IsDone check and Done are not done atomically. This is why another thread can call Commit around the same instant when the first one has already started executing the ops. I think this could be a problem in the current inmem txchunkstore as well. Wdyt?